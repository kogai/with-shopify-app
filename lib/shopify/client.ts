import querystring from 'querystring'
import crypto from 'crypto'
import safeCompare from 'safe-compare'
import fetch from 'isomorphic-fetch'
import { GraphQLClient } from 'graphql-request'
import { getSdk } from './admin-api' // THIS FILE IS THE GENERATED FILE
import { Result, createOk, createErr } from '../result'

type Config = {
  clientSecret: string
  clientID: string
  version: string
  appURL: string
  scope: string
}

export type VerifyPayload = {
  hmac: string
  [key: string]: string
}

export type AccessTokenPayload = {
  code: string
  shop: string
}

export type AccessTokenResponse = {
  access_token: string
}

export type StorefrontAccessTokenPayload = {
  accessToken: string
  shop: string
  title: string
}

export type StorefrontAccessTokenResponse = {
  storefront_access_token: {
    access_token: string
  }
}

export type AuthorizationURLResponse = {
  shop: string
  state: string
}

export type AppIframeURLResponse = {
  accessToken: string
  shop: string
}

interface ShopifyClient {
  authorizationURL(payload: AuthorizationURLResponse): Result<string, Error>
  appIframeURL(payload: AppIframeURLResponse): Promise<Result<string, Error>>
  verifyMessage(payload: VerifyPayload): boolean
  accessTokenCreate(
    payload: AccessTokenPayload
  ): Promise<Result<AccessTokenResponse, Error>>
  storefrontAccessToken(
    payload: StorefrontAccessTokenPayload
  ): Promise<Result<StorefrontAccessTokenResponse, Error>>
}

const verifyMessage = (config: Config): ShopifyClient['verifyMessage'] => ({
  hmac,
  ...rest
}) => {
  const orderedMap = Object.keys(rest)
    .sort((value1, value2) => value1.localeCompare(value2))
    .reduce((accum, key) => {
      accum[key] = rest[key]
      return accum
    }, {} as VerifyPayload)
  const message = querystring.stringify(orderedMap)

  const generatedHash = crypto
    .createHmac('sha256', config.clientSecret)
    .update(message)
    .digest('hex')

  return safeCompare(generatedHash, hmac)
}

const accessTokenCreate = ({
  clientID,
  clientSecret,
}: Config): ShopifyClient['accessTokenCreate'] => async ({ code, shop }) => {
  const body = querystring.stringify({
    code,
    client_id: clientID,
    client_secret: clientSecret,
  })
  // NOTE: OAuth API is unversioned.
  // See https://shopify.dev/concepts/about-apis/versioning
  const accessTokenRes = await fetch(
    `https://${shop}/admin/oauth/access_token`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': Buffer.byteLength(body).toString(),
      },
      body: body,
    }
  )
  if (!accessTokenRes.ok) {
    const err = await accessTokenRes.text()
    return createErr(new Error(err))
  }
  return createOk(await accessTokenRes.json())
}

const storefrontAccessToken = ({
  version,
}: Config): ShopifyClient['storefrontAccessToken'] => async ({
  title,
  shop,
  accessToken,
}) => {
  const storeFrontAccessTokenRes = await fetch(
    `https://${shop}/admin/api/${version}/storefront_access_tokens.json`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Shopify-Access-Token': accessToken,
      },
      body: JSON.stringify({
        storefront_access_token: {
          title,
        },
      }),
    }
  )
  if (!storeFrontAccessTokenRes.ok) {
    const err = await storeFrontAccessTokenRes.text()
    return createErr(new Error(err))
  }
  return createOk(await storeFrontAccessTokenRes.json())
}

const authorizationURL = ({
  appURL,
  clientID,
  scope,
}: Config): ShopifyClient['authorizationURL'] => ({ shop, state }) => {
  const DEFAULT_MYSHOPIFY_DOMAIN = 'myshopify.com'
  const shopRegex = new RegExp(
    `^[a-z0-9][a-z0-9\\-]*[a-z0-9]\\.${DEFAULT_MYSHOPIFY_DOMAIN}$`,
    'i'
  )
  if (!shopRegex.test(shop)) {
    return createErr(new Error('invalid shop name'))
  }
  const message = querystring.stringify({
    client_id: clientID,
    redirect_uri: `${appURL}/api/done`,
    scope,
    state,
  })
  return createOk(`https://${shop}/admin/oauth/authorize?${message}`)
}

const appIframeURL = ({
  version,
}: Config): ShopifyClient['appIframeURL'] => async ({ shop, accessToken }) => {
  const client = new GraphQLClient(
    `https://${shop}/admin/api/${version}/graphql.json`,
    {
      headers: {
        'Content-Type': 'application/json',
        'X-Shopify-Access-Token': accessToken,
      },
    }
  )
  try {
    const { app } = await getSdk(client).appHandle()
    if (app?.handle == null) {
      return createErr(new Error(`Expected value cant get. ${app}`))
    }
    return createOk(`https://${shop}/admin/apps/${app.handle}`)
  } catch (error) {
    return createErr(error)
  }
}

export const createClient = (config_: Config): ShopifyClient => {
  const appURL = /\/$/.test(config_.appURL)
    ? config_.appURL.substring(0, config_.appURL.length - 1)
    : config_.appURL
  const config = {
    ...config_,
    appURL,
  }
  return {
    verifyMessage: verifyMessage(config),
    accessTokenCreate: accessTokenCreate(config),
    storefrontAccessToken: storefrontAccessToken(config),
    authorizationURL: authorizationURL(config),
    appIframeURL: appIframeURL(config),
  }
}
