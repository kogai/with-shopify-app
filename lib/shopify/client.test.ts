import fs from 'fs'
import crypto from 'crypto'
import nock from 'nock'
import querystring from 'querystring'
import { createClient } from './client'

// NOTE: Test helper that produce hmac signed string.
const signHmac = () =>
  crypto
    .createHmac('sha256', 'test_api_secret')
    .update(
      querystring.stringify({
        code: '035fa8d75e6c5628d2406f5918e81529',
        shop: 'mytest.myshopify.com',
        state: '4d358df9cfa5826d1ac8cc8b9ffc18ca',
        timestamp: '1592874758',
      })
    )
    .digest('hex')

describe('lib/shopify/client', () => {
  beforeAll(() => nock.disableNetConnect())
  afterEach(() => nock.cleanAll())

  const client = createClient({
    version: '2099-00',
    clientID: 'test_api_key',
    clientSecret: 'test_api_secret',
    appURL: 'https://example.com',
    scope: 'administration',
  })

  describe('verify message', () => {
    it('signHmac utility', () => {
      expect(signHmac()).toBe(
        '42cca3f99694a606fcd0cb1269c804d723828f5d6abef2eece1bef9e02a11d96'
      )
    })
    ;[
      {
        title: 'ok',
        hmac:
          '42cca3f99694a606fcd0cb1269c804d723828f5d6abef2eece1bef9e02a11d96',
        expected: true,
      },
      {
        title: 'ng',
        hmac:
          'aaaca3f99694a606fcd0cb1269c804d723828f5d6abef2eece1bef9e02a11d96',
        expected: false,
      },
    ].forEach(({ title, hmac, expected }) => {
      it(title, () => {
        const ok = client.verifyMessage({
          hmac,
          code: '035fa8d75e6c5628d2406f5918e81529',
          shop: 'mytest.myshopify.com',
          state: '4d358df9cfa5826d1ac8cc8b9ffc18ca',
          timestamp: '1592874758',
        })
        expect(ok).toBe(expected)
      })
    })
  })

  describe('access-token', () => {
    it('ok', async () => {
      nock('https://mytest.myshopify.com', {
        reqheaders: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Content-Length': '90',
        },
      })
        .post(
          '/admin/oauth/access_token',
          'code=035fa8d75e6c5628d2406f5918e81529&client_id=test_api_key&client_secret=test_api_secret'
        )
        .reply(200, {
          access_token: 'abc',
        })
      const result = await client.accessTokenCreate({
        code: '035fa8d75e6c5628d2406f5918e81529',
        shop: 'mytest.myshopify.com',
      })
      expect(result).toMatchInlineSnapshot(`
        Object {
          "err": undefined,
          "ok": true,
          "val": Object {
            "access_token": "abc",
          },
        }
      `)
    })

    it('ng', async () => {
      nock('https://mytest.myshopify.com', {
        reqheaders: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Content-Length': '90',
        },
      })
        .post(
          '/admin/oauth/access_token',
          'code=035fa8d75e6c5628d2406f5918e81529&client_id=test_api_key&client_secret=test_api_secret'
        )
        .reply(500, {
          message: 'internal server error',
        })
      const result = await client.accessTokenCreate({
        code: '035fa8d75e6c5628d2406f5918e81529',
        shop: 'mytest.myshopify.com',
      })
      expect(result).toMatchInlineSnapshot(`
        Object {
          "err": [Error: {"message":"internal server error"}],
          "ok": false,
          "val": undefined,
        }
      `)
    })
  })

  describe('storefront-access-token', () => {
    it('ok', async () => {
      nock('https://mytest.myshopify.com', {
        reqheaders: {
          'Content-Type': 'application/json',
          'X-Shopify-Access-Token': 'abc',
        },
      })
        .post(
          '/admin/api/2099-00/storefront_access_tokens.json',
          JSON.stringify({
            storefront_access_token: {
              title: 'label-of-token',
            },
          })
        )
        .reply(200, {
          storefront_access_token: {
            access_token: 'def',
          },
        })
      const result = await client.storefrontAccessToken({
        accessToken: 'abc',
        title: 'label-of-token',
        shop: 'mytest.myshopify.com',
      })
      expect(result).toMatchInlineSnapshot(`
        Object {
          "err": undefined,
          "ok": true,
          "val": Object {
            "storefront_access_token": Object {
              "access_token": "def",
            },
          },
        }
      `)
    })

    it('ng', async () => {
      nock('https://mytest.myshopify.com', {
        reqheaders: {
          'Content-Type': 'application/json',
          'X-Shopify-Access-Token': 'abc',
        },
      })
        .post(
          '/admin/api/2099-00/storefront_access_tokens.json',
          JSON.stringify({
            storefront_access_token: {
              title: 'label-of-token',
            },
          })
        )
        .reply(500, {
          message: 'internal server error',
        })
      const result = await client.storefrontAccessToken({
        accessToken: 'abc',
        title: 'label-of-token',
        shop: 'mytest.myshopify.com',
      })
      expect(result).toMatchInlineSnapshot(`
        Object {
          "err": [Error: {"message":"internal server error"}],
          "ok": false,
          "val": undefined,
        }
      `)
    })
  })

  describe('authorization URL', () => {
    it('ok', async () => {
      const result = client.authorizationURL({
        shop: 'mytest.myshopify.com',
        state: 'abc',
      })
      expect(result).toMatchInlineSnapshot(`
        Object {
          "err": undefined,
          "ok": true,
          "val": "https://mytest.myshopify.com/admin/oauth/authorize?client_id=test_api_key&redirect_uri=https%3A%2F%2Fexample.com%2Fapi%2Fdone&scope=administration&state=abc",
        }
      `)
    })

    it('ng', async () => {
      const result = client.authorizationURL({
        shop: 'mytest.bigcommerce.com',
        state: 'abc',
      })
      expect(result).toMatchInlineSnapshot(`
        Object {
          "err": [Error: invalid shop name],
          "ok": false,
          "val": undefined,
        }
      `)
    })

    it('strip uselss slash', async () => {
      const client = createClient({
        version: '2099-00',
        clientID: 'test_api_key',
        clientSecret: 'test_api_secret',
        appURL: 'https://example.com/',
        scope: 'administration',
      })
      const result = client.authorizationURL({
        shop: 'mytest.myshopify.com',
        state: 'abc',
      })
      expect(result).toMatchInlineSnapshot(`
        Object {
          "err": undefined,
          "ok": true,
          "val": "https://mytest.myshopify.com/admin/oauth/authorize?client_id=test_api_key&redirect_uri=https%3A%2F%2Fexample.com%2Fapi%2Fdone&scope=administration&state=abc",
        }
      `)
    })
  })

  describe('app ifram url', () => {
    ;[
      {
        title: 'ok',
        statusCode: 200,
        response: {
          data: {
            app: {
              handle: 'myapp',
            },
          },
        },
      },
      {
        title: 'ng(json)',
        statusCode: 500,
        response: {
          message: 'internal server error',
        },
      },
      {
        title: 'ng(text)',
        statusCode: 500,
        response: 'internal server error',
      },
    ].forEach(({ title, statusCode, response }) => {
      it(title, async () => {
        nock('https://mytest.myshopify.com', {
          reqheaders: {
            'content-type': 'application/json',
            'x-shopify-access-token': 'abc',
          },
        })
          .post(
            '/admin/api/2099-00/graphql.json',
            JSON.stringify({
              query: fs.readFileSync('./lib/shopify/appHandle.gql').toString(),
            })
          )
          .reply(statusCode, response)

        const result = await client.appIframeURL({
          accessToken: 'abc',
          shop: 'mytest.myshopify.com',
        })
        expect(result).toMatchSnapshot()
      })
    })
  })
})
