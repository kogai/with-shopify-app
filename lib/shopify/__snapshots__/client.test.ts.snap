// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/shopify/client app ifram url ng(json) 1`] = `
Object {
  "err": [Error: GraphQL Error (Code: 500): {"response":{"message":"internal server error","status":500},"request":{"query":"query appHandle {\\n  app {\\n    handle\\n  }\\n}\\n"}}],
  "ok": false,
  "val": undefined,
}
`;

exports[`lib/shopify/client app ifram url ng(text) 1`] = `
Object {
  "err": [Error: GraphQL Error (Code: 500): {"response":{"error":"internal server error","status":500},"request":{"query":"query appHandle {\\n  app {\\n    handle\\n  }\\n}\\n"}}],
  "ok": false,
  "val": undefined,
}
`;

exports[`lib/shopify/client app ifram url ok 1`] = `
Object {
  "err": undefined,
  "ok": true,
  "val": "https://mytest.myshopify.com/admin/apps/myapp",
}
`;
